### ISSUE TITLE
function undefined arguments - how to test for undefined? is this a bug?

### ISSUE THREAD

**Issue by hrvoj3e** [fish-shell/fish-shell#5252]
How to test if named arguments are provided to function or not?

Or should I rather use `$argv`, `$argv[1]`, ... ?

E.g. `test -z $two` and `test -n $two` are both satisfied.

```shell
function bug1 -a one two -d "testing"
    if test -z $two
        echo "#1 two invalid"
    end
    if test -n $two
        echo "#1 two valid"
    end

    if test -z "$two"
        echo "#2 two invalid"
    end
    if test -n "$two"
        echo "#2 two valid"
    end
end
```

```shell
source .config/fish/functions/bug1.fish; bug1 111
#1 two invalid
#1 two valid
#2 two invalid
```

> 
> fish, version 2.7.1
> xterm-256color
> Linux adax-pc 4.18.12-1-MANJARO #1 SMP PREEMPT Fri Oct 5 16:06:16 UTC 2018 x86_64 GNU/Linux

**Comment by faho** [comment]
I'm just gonna copy an old stackoverflow answer of mine:

- set -q var (note the missing "$" - this uses the variable name) can be used to check if a variable has been set.

- set -q var[1] can be used to check whether the first element of a variable has been assigned (i.e. whether it is non-empty as a list).

- test -n "$var" [fn0] (or [ -n "$var" ]) can be used to check whether a variable expands to a non-empty string (and test -z is the inverse - true if it is empty).

These will be true/false in slightly different circumstances.

When no set var has been performed at all (and it has not been inherited from the parent process), set -q var, set -q var[1] and test -n "$var" will be false, test -z "$var" will be true.

When something like set var has been done (without any additional arguments), set -q var will be true, set -q var[1] will be false.

When something like set var "" has been done, both set versions will be true.

When something like set var "somestring" (or even set var "" "" [fn1]) has been done, the sets will be true and test -z "$var" will be false.

[fn0]: You never want to use test (or [) without quoting the variable. One particularly egregious example is that test -n $var will return true both if the variable contains something and if it is list-empty/unset (no set at all or set var without arguments). This is because fish's test is one of the few parts that follow POSIX, and that demands that test with any one argument be true. Also it does not handle lists properly - test -n $var will have weird results if var has more than one element.

[fn1]: This is because a list will be expanded as a string by joining the elements with spaces, so the list consisting of two empty strings will expand to " " - one space. Since that isn't empty, test -z returns false.

TL;DR: Never use `test` without quotes, `set -q var` and `set -q var[1]` are nice.

**Comment by hrvoj3e** [comment]
Thank you for the explanation - it is much clearer now.
Your stack post is helping a lot too (https://stackoverflow.com/questions/47743015/fish-shell-how-to-check-if-a-variable-is-set-empty/47743269#47743269).
I based my search/testing on "named arguments" - but they are obviously normal variables.
